tests/0105-transactions_mock.c

```diff
@@ -2866,6 +2866,96 @@ static void do_test_disconnected_group_coord(rd_bool_t switch_coord) {
+        const char *topic            = "test";
+        const char *transactional_id = "txnid";
+        int msgcnt                   = 1;
+        int remains                  = 0;
+
+        SUB_TEST_QUICK("Test coordinator NULL shouldn't cause "
+            "a fatal assertion "
+            "when transactional producer reconnects");
+
+        test_curr->is_fatal_cb = error_is_fatal_cb;
+        test_curr->exp_dr_err  = RD_KAFKA_RESP_ERR__MSG_TIMED_OUT;
+        allowed_error          = RD_KAFKA_RESP_ERR__TRANSPORT;
+
+        rk = create_txn_producer(&mcluster, transactional_id, 1, "debug", "all", NULL);
+        rkt = test_create_topic_object(rk, topic, "message.timeout.ms", "100", NULL);
+
+        /* Broker down is not a test-failing error */
+        allowed_error          = RD_KAFKA_RESP_ERR__TRANSPORT;
+        test_curr->is_fatal_cb = error_is_fatal_cb;
+
+        err = rd_kafka_mock_topic_create(mcluster, topic, 1, 3);
+        TEST_ASSERT(!err, "Failed to create topic: %s", rd_kafka_err2str(err));
+
+        rd_kafka_mock_coordinator_set(mcluster, "transaction", transactional_id,
+                                      coord_id);
+        rd_kafka_mock_partition_set_leader(mcluster, topic, 0, coord_id);
+
+        /* Start transactioning */
+        TEST_SAY("Starting transaction\n");
+        TEST_CALL_ERROR__(rd_kafka_init_transactions(rk, 5000));
+        TEST_CALL_ERROR__(rd_kafka_begin_transaction(rk));
+
+        test_produce_msgs_nowait(rk, rkt, 0, RD_KAFKA_PARTITION_UA, 0,
+                                 msgcnt, NULL, 0, 0, &remains);
+
+
+        TEST_SAY("Bringing down coordinator %" PRId32 "\n", coord_id);
+        rd_kafka_mock_broker_set_down(mcluster, coord_id);
+
+        rd_kafka_mock_broker_push_request_error_rtts(
+            mcluster, coord_id, RD_KAFKAP_FindCoordinator, 1,
+                RD_KAFKA_RESP_ERR_COORDINATOR_NOT_AVAILABLE, 10);
+
+        /* This value is linked to message.timeout.ms, needs enough time
+         * so message time out and aren't purged */
+        rd_kafka_flush(rk, 500);
+        /* Needs a specific timeout so FindCoordinatorRequest is the first thing that
+         * is done when the coordinator comes up */
+        rd_usleep(605 * 1000, NULL);
+
+        TEST_SAY("Bringing up coordinator %" PRId32 "\n", coord_id);
+            rd_kafka_mock_broker_set_up(mcluster, coord_id);
+
+        error = rd_kafka_commit_transaction(rk, -1);
+
+        TEST_ASSERT(remains == 0, "%d message(s) were not produced\n", remains);
+        TEST_ASSERT(error != NULL, "Expected commit_transaction() to fail");
+
+        TEST_SAY("commit_transaction() failed (expectedly): %s\n",
+                 rd_kafka_error_string(error));
+
+        /* Needs to wait some time before closing to make sure it doesn't go
+         * into TERMINATING state before error is triggered */
+        rd_usleep(1400 * 1000, NULL);
+
+        rd_kafka_topic_destroy(rkt);
+        rd_kafka_error_destroy(error);
```

Comment:
it is better to destroy temporary error objects after their last use, so on line 2942.